File: app/src/main/java/com/callmangement/ui/reports/ExpenseDetailsActivity.java
package com.callmangement.ui.reports;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.callmangement.network.APIService;
import com.callmangement.network.RetrofitInstance;
import com.callmangement.R;
import com.callmangement.custom.CustomActivity;
import com.callmangement.databinding.ActivityExpenseDetailsBinding;
import com.callmangement.model.expense.ModelExpensesList;
import com.callmangement.ui.home.ZoomInZoomOutActivity;
import com.callmangement.utils.Constants;
import com.callmangement.utils.PrefManager;
import org.json.JSONObject;

import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class ExpenseDetailsActivity extends CustomActivity implements View.OnClickListener {
    private ActivityExpenseDetailsBinding binding;
    private ModelExpensesList model;
    private PrefManager prefManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityExpenseDetailsBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        initView();
    }

    private void initView() {
        prefManager = new PrefManager(mContext);
        binding.actionBar.ivBack.setVisibility(View.VISIBLE);
        binding.actionBar.ivThreeDot.setVisibility(View.GONE);
        binding.actionBar.layoutLanguage.setVisibility(View.GONE);
        binding.actionBar.textToolbarTitle.setText(getResources().getString(R.string.expense_detail));
        setUpOnClickListener();
        getIntentData();
        setUpData();
    }

    private void setUpOnClickListener() {
        binding.actionBar.ivBack.setOnClickListener(this);
        binding.ivChallanImage.setOnClickListener(this);
        binding.buttonComplete.setOnClickListener(this);
    }

    private void getIntentData() {
        model = (ModelExpensesList) getIntent().getSerializableExtra("param");
    }

    private void setUpData(){
        binding.inputName.setText(model.getSeName());
        binding.inputDistrict.setText(model.getDistrict());

        binding.inputDocketno.setText(model.docketNo);
        binding.inputcouriername.setText(model.courierName);

        binding.inputRemark.setText(model.getRemark());
        binding.inputRemark.setText(model.getRemark());
        binding.inputTotalExpenseAmount.setText(String.valueOf(model.totalExAmount));
        binding.inputExpenseDate.setText(model.getCreatedOnStr());

        if (prefManager.getUSER_TYPE_ID().equals("1") && prefManager.getUSER_TYPE().equalsIgnoreCase("Admin")) {
            binding.buttonComplete.setVisibility(View.VISIBLE);
        } else if (prefManager.getUSER_TYPE_ID().equals("2") && prefManager.getUSER_TYPE().equalsIgnoreCase("Manager")) {
            binding.buttonComplete.setVisibility(View.VISIBLE);
        } else {
            binding.buttonComplete.setVisibility(View.GONE);
        }

        if (model.getCompletedOnStr().isEmpty()){
            binding.expenseCompletedDateLay.setVisibility(View.GONE);
        } else {
            binding.expenseCompletedDateLay.setVisibility(View.VISIBLE);
            binding.inputExpenseCompletedDate.setText(model.getCompletedOnStr());
        }

        if (model.getExpenseStatusID() == 2) {
            binding.buttonComplete.setVisibility(View.GONE);
        }

        Glide.with(mContext)
                .load(Constants.API_BASE_URL+ model.filePath)
                .placeholder(R.drawable.image_not_fount)
                .diskCacheStrategy(DiskCacheStrategy.NONE)
                .into(binding.ivChallanImage);

    }

    private void completeExpense() {
        if (Constants.isNetworkAvailable(mContext)) {
            showProgress();
            APIService service = RetrofitInstance.getRetrofitInstance().create(APIService.class);
            Call<ResponseBody> call = service.completeExpenses(prefManager.getUSER_Id(), String.valueOf(model.getExpenseId()));
            call.enqueue(new Callback<ResponseBody>() {
                @Override
                public void onResponse(@NonNull Call<ResponseBody> call, @NonNull Response<ResponseBody> response) {
                    hideProgress();
                    try {
                        if (response.isSuccessful()) {
                            if (response.code() == 200) {
                                if (response.body() != null) {
                                    String responseStr = response.body().string();
                                    JSONObject jsonObject = new JSONObject(responseStr);
                                    String status = jsonObject.optString("status");
                                    String message = jsonObject.optString("message");
                                    if (status.equals("200")) {
                                        Toast.makeText(mContext, message, Toast.LENGTH_SHORT).show();
                                        onBackPressed();
                                    } else {
                                        makeToast(message);
                                    }
                                }
                            }
                        } else {
                            makeToast(getResources().getString(R.string.error));
                        }
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }

                @Override
                public void onFailure(@NonNull Call<ResponseBody> call, @NonNull Throwable t) {
                    hideProgress();
                    makeToast(getResources().getString(R.string.error_message));
                }
            });
        } else
            makeToast(getResources().getString(R.string.no_internet_connection));
    }

    @Override
    public void onClick(View view) {
        int id = view.getId();
        if (id == R.id.iv_back){
            onBackPressed();
        } else if (id == R.id.ivChallanImage){
            startActivity(new Intent(mContext, ZoomInZoomOutActivity.class).putExtra("image", Constants.API_BASE_URL+ model.filePath));
        } else if (id == R.id.buttonComplete) {
            AlertDialog.Builder builder = new AlertDialog.Builder(mContext);
            builder.setMessage(getResources().getString(R.string.complete_expense_dialog_msg))
                    .setCancelable(false)
                    .setPositiveButton(getResources().getString(R.string.ok), (dialog, ids) -> {
                        dialog.cancel();
                        completeExpense();
                    })
                    .setNegativeButton(getResources().getString(R.string.cancel), (dialog, ids) -> dialog.cancel());
            AlertDialog alert = builder.create();
            alert.setTitle(getResources().getString(R.string.alert));
            alert.show();
        }
    }
}

File: app/src/main/java/com/callmangement/ui/reports/ExpenseDetailsActivity.kt
package com.callmangement.ui.reports

import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.callmangement.Network.APIService
import com.callmangement.Network.RetrofitInstance
import com.callmangement.R
import com.callmangement.custom.CustomActivity
import com.callmangement.databinding.ActivityExpenseDetailsBinding
import com.callmangement.model.expense.ModelExpensesList
import com.callmangement.ui.home.ZoomInZoomOutActivity
import com.callmangement.utils.Constants
import com.callmangement.utils.PrefManager
import okhttp3.ResponseBody
import org.json.JSONObject
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class ExpenseDetailsActivity : CustomActivity(), View.OnClickListener {
    private var binding: ActivityExpenseDetailsBinding? = null
    private var model: ModelExpensesList? = null
    private var prefManager: PrefManager? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityExpenseDetailsBinding.inflate(layoutInflater)
        setContentView(binding!!.root)
        initView()
    }

    private fun initView() {
        prefManager = PrefManager(mContext)
        binding!!.actionBar.ivBack.visibility = View.VISIBLE
        binding!!.actionBar.ivThreeDot.visibility = View.GONE
        binding!!.actionBar.layoutLanguage.visibility = View.GONE
        binding!!.actionBar.textToolbarTitle.text = resources.getString(R.string.expense_detail)
        setUpOnClickListener()
        intentData
        setUpData()
    }

    private fun setUpOnClickListener() {
        binding!!.actionBar.ivBack.setOnClickListener(this)
        binding!!.ivChallanImage.setOnClickListener(this)
        binding!!.buttonComplete.setOnClickListener(this)
    }

    private val intentData: Unit
        get() {
            model = intent.getSerializableExtra("param") as ModelExpensesList?
        }

    private fun setUpData() {
        binding!!.inputName.setText(model!!.seName)
        binding!!.inputDistrict.setText(model!!.district)

        binding!!.inputDocketno.setText(model!!.docketNo)
        binding!!.inputcouriername.setText(model!!.courierName)

        binding!!.inputRemark.setText(model!!.remark)
        binding!!.inputRemark.setText(model!!.remark)
        binding!!.inputTotalExpenseAmount.setText(model!!.totalExAmount.toString())
        binding!!.inputExpenseDate.setText(model!!.createdOnStr)

        if (prefManager!!.useR_TYPE_ID == "1" && prefManager!!.useR_TYPE.equals(
                "Admin",
                ignoreCase = true
            )
        ) {
            binding!!.buttonComplete.visibility = View.VISIBLE
        } else if (prefManager!!.useR_TYPE_ID == "2" && prefManager!!.useR_TYPE.equals(
                "Manager",
                ignoreCase = true
            )
        ) {
            binding!!.buttonComplete.visibility = View.VISIBLE
        } else {
            binding!!.buttonComplete.visibility = View.GONE
        }

        if (model!!.completedOnStr.isEmpty()) {
            binding!!.expenseCompletedDateLay.visibility = View.GONE
        } else {
            binding!!.expenseCompletedDateLay.visibility = View.VISIBLE
            binding!!.inputExpenseCompletedDate.setText(model!!.completedOnStr)
        }

        if (model!!.expenseStatusID == 2) {
            binding!!.buttonComplete.visibility = View.GONE
        }

        Glide.with(mContext)
            .load(Constants.API_BASE_URL + model!!.filePath)
            .placeholder(R.drawable.image_not_fount)
            .diskCacheStrategy(DiskCacheStrategy.NONE)
            .into(binding!!.ivChallanImage)
    }

    private fun completeExpense() {
        if (Constants.isNetworkAvailable(mContext)) {
            showProgress()
            val service = RetrofitInstance.getRetrofitInstance().create(
                APIService::class.java
            )
            val call = service.completeExpenses(prefManager!!.useR_Id, model!!.expenseId.toString())
            call.enqueue(object : Callback<ResponseBody?> {
                override fun onResponse(
                    call: Call<ResponseBody?>,
                    response: Response<ResponseBody?>
                ) {
                    hideProgress()
                    try {
                        if (response.isSuccessful) {
                            if (response.code() == 200) {
                                if (response.body() != null) {
                                    val responseStr = response.body()!!.string()
                                    val jsonObject = JSONObject(responseStr)
                                    val status = jsonObject.optString("status")
                                    val message = jsonObject.optString("message")
                                    if (status == "200") {
                                        Toast.makeText(mContext, message, Toast.LENGTH_SHORT).show()
                                        onBackPressed()
                                    } else {
                                        makeToast(message)
                                    }
                                }
                            }
                        } else {
                            makeToast(resources.getString(R.string.error))
                        }
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }
                }

                override fun onFailure(call: Call<ResponseBody?>, t: Throwable) {
                    hideProgress()
                    makeToast(resources.getString(R.string.error_message))
                }
            })
        } else makeToast(resources.getString(R.string.no_internet_connection))
    }

    override fun onClick(view: View) {
        val id = view.id
        if (id == R.id.iv_back) {
            onBackPressed()
        } else if (id == R.id.ivChallanImage) {
            startActivity(
                Intent(mContext, ZoomInZoomOutActivity::class.java).putExtra(
                    "image",
                    Constants.API_BASE_URL + model!!.filePath
                )
            )
        } else if (id == R.id.buttonComplete) {
            val builder = AlertDialog.Builder(mContext)
            builder.setMessage(resources.getString(R.string.complete_expense_dialog_msg))
                .setCancelable(false)
                .setPositiveButton(resources.getString(R.string.ok)) { dialog: DialogInterface, ids: Int ->
                    dialog.cancel()
                    completeExpense()
                }
                .setNegativeButton(
                    resources.getString(R.string.cancel)
                ) { dialog: DialogInterface, ids: Int -> dialog.cancel() }
            val alert = builder.create()
            alert.setTitle(resources.getString(R.string.alert))
            alert.show()
        }
    }
}